### DISCUSSION TITLE
Elastic - Transform requires the transform node role for at least 1 node

### DISCUSSION THREAD

**Discussion by Mav1814** [Security-Onion-Solutions/securityonion#11185] [2.4] [answered]
### Version

2.4.10

### Installation Method

Security Onion ISO image

### Description

configuration

### Installation Type

Distributed

### Location

on-prem with Internet access

### Hardware Specs

Exceeds minimum requirements

### Network Traffic Collection

tap

### Network Traffic Speeds

1Gbps to 10Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

No, there are no failures

### Logs

No, there are no additional clues

### Detail

Hello, 

How can i fix?

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/b2af28cb-9c11-46dd-8fc1-3a5f26da61d9)

I know that is a problem with transforms, but I could not fix it. I saw here ##11084 to add a function for the transform, but I didn't understand it.

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by Mav1814** [comment]
Any suggestion?

**Reply by weslambert** [reply]
You should be able to follow the steps here to resolve your issue:

https://github.com/Security-Onion-Solutions/securityonion/discussions/11084#discussioncomment-6826251

**Reply by Mav1814** [reply]
Hi, 
As i said before 
"I saw here #https://github.com/Security-Onion-Solutions/securityonion/discussions/11084 to add a function for the transform, but I didn't understand it."
I've tried to fix on this past days but no success..

**Reply by dougburks** [reply]
> I saw here #https://github.com/Security-Onion-Solutions/securityonion/discussions/11084 to add a function for the transform, but I didn't understand it.

If you tell us which part of this you don't understand, then perhaps we can help you understand it.

From https://github.com/Security-Onion-Solutions/securityonion/discussions/11084#discussioncomment-6826251:

```
Remove the Elastic Defend integration by navigating to Elastic Fleet -> Agent Policies -> endpoints-initial -> Actions -> Delete integration. Then run the following on the Manager:

sudo rm -f /opt/so/state/eaintegrations.txt
sudo so-elastic-fleet-integration-policy-load

If you are running a distributed grid with a Manager or ManagerSearch, you may also need to add the following role to Elasticsearch:
'transform'
sudo vi +24 /opt/so/saltstack/default/salt/elasticsearch/config.map.jinja 
so-elasticsearch-restart
```

**Reply by Mav1814** [reply]
Hi, @dougburks 
thank you for everything in advance. 
 
I've been trying to add the "transform" role to elasticsearch as seen below, but I'm still having the same problem as shown in the image above.
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/c4855264-512a-4054-8162-65da4c532ce2)

"status_exception Root causes: status_exception: Transform requires the transform node role for at least 1 node, found no transform nodes"
And the status are "stopped"
The elasticsearch restart without a issue. I dont know what im doing wrong.

**Reply by Mav1814** [reply]
I read i don't remember were, that it would be a "transform" by "remote_cluster_client". So i added two..

**Reply by dougburks** [reply]
So what does your edited version of the file look like?

**Reply by Mav1814** [reply]
It looks like the image that i put above. 
i did "sudo vi +24 /opt/so/saltstack/default/salt/elasticsearch/config.map.jinja "
Then i added 
` {% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.update({'roles': ['master','data','remote_cluster_client']}) %}
      {% if HIGHLANDER %}
          {% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.roles.extend(['ml', 'transform']) %}
      {% endif %}
        {% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.update({'roles': ['master', 'data_hot', 'remote_cluster_client']}) %}
      {% if HIGHLANDER %}
          {% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.roles.extend(['ml', 'transform']) %}
      {% endif %}
    {% endif %}
`

**Reply by Mav1814** [reply]
And i still get this result, even after restart elasticsearch

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/af52b815-ed54-45b9-b343-257f1a8de12a)

**Reply by dougburks** [reply]
What you [pasted above](https://github.com/Security-Onion-Solutions/securityonion/discussions/11185#discussioncomment-6978740) does not look correct.

In the current release, the default version of `/opt/so/saltstack/default/salt/elasticsearch/config.map.jinja` should look like this on line 24:
```
{% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.update({'roles': ['master','data','remote_cluster_client']}) %}
```

If you add `transform` to the roles, then it should look like this:
```
{% do ELASTICSEARCHDEFAULTS.elasticsearch.config.node.update({'roles': ['master','data','remote_cluster_client','transform']}) %}
```

Please double-check that you updated line 24 correctly.

**Reply by Mav1814** [reply]
I will try it right now.!! 

I will post here the results! Much tnx @dougburks u guys are amazing!

**Reply by Mav1814** [reply]
I made the change that you said. 

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/a6e48c51-6404-477b-b618-840c7ce480fe)

I restarted the Elasticsearch, is working, but I still get the error.. Should i do anything else? Or ?

**Comment by defensivedepth** [comment]
hey there @Mav1814... For our context, what are you attempting to do in the Transforms page?

Also, do you have any 8.7 Elastic Agents still deployed?

**Reply by Mav1814** [reply]
Hey @defensivedepth 
First I'm trying to get the transforms working. I want to understand how useful they can be in real-time data analysis scenarios or even to create data visualizations and aggregations to support insights and decision-making. It also allows data to be reformatted into a target index to facilitate analysis and consultation.

I don't any Agent with the 8.7, they are all with 8.8.2.

**Reply by defensivedepth** [reply]
What version of Security Onion did you originally install with? What specific nodes are you running?

Do you see that error for the 8.8.0 indices?

**Reply by Mav1814** [reply]
I installed version 2.4.3 Beta 4 [20230711] and then kept updating to the current version 2.4.10 Hotfix [20230821]. 
My distributed system consists of:  
- 1 Manager
- 1 Fleet Server
- 2 Search nodes
- 2 Forwards nodes on different networks
- 1 Receiver
- 1 IDH

I'm still in the testing phase, trying to understand how the architecture of the distributed system works. The goal is to implement the SO effectively in my organization, but we still have some doubts about how it works, especially in terms of logs and dashboards. This is because we've noticed that most of the fields in the logs, both from the SO dashboard/hunts and from the kibana side, aren't filled in. It's probably for some reason of data extraction, we're trying to figure it out. Hence the attempt to use transforms. 

"Do you see that error for the 8.8.0 indices?"
From the transforms? If i go to preview on the 8.8.0 indices of the transforms yes, the error is there " An error occurred loading the index data. Please ensure your query syntax is valid.
Bad Request: [[status_exception] Transform requires the transform node role for at least 1 node, found no transform nodes]: Transform requires the transform node role for at least 1 node, found no transform nodes"

**Reply by defensivedepth** [reply]
A couple notes:

- Feel free to open a new Discussion for missing data, and we can troubleshoot that. It should work out of the box without Transforms.

- From the Known Issues for 2.4.3 Beta 4 (https://blog.securityonion.net/2023/07/security-onion-24-beta-4-release-now.html):
`Upgrades from this 2.4 Beta release to anything else will not be supported. Starting in RC2 we will support soup to upgrade 2.4 grids.`

Upgrading from Beta 4 is not supported, and is most likely the root issue of what you are seeing. I would suggest a fresh install from GA.

**Reply by Mav1814** [reply]
@defensivedepth  my mistake, i have 2 "isos", i confirmed that i installed from 2.4.4. 

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/36708408/d575e413-b062-4c3a-9cf1-53d24a296f80)

I tend to be careful and read the documentation well. And I thought the beta was strange, so I went to check. Because I remember using SOUP. So I could never have installed it from beta. Otherwise I wouldn't even be on version 2.4.10. It was my lack of attention in my reply. I'm sorry for the wrong info.

**Reply by Mav1814** [reply]
Yeh i will open a new discussion. :) 

For the time being, its better to do a "fresh" from the GA?

**Reply by Mav1814** [reply, marked as answer]
I decided to make a fresh install. So i will close this discussion. Thank you :)
